TACC/Core-CMS
# feat(v3.6.0): tup 230 make form plugin work

## Overview

Form plugin: update, configure, style*, document.

<sub>* The styles only use inspiration from [new TUP login form](https://xd.adobe.com/view/1bebd965-c072-4fac-8727-ba61c5ca3baf-4eb8/screen/462fd224-c85d-45df-9796-9d21eb7d3b1a/specs/). Design will later give standard CMS form styles (and later CMS-specific button styles).</sub>

## Related

- [TUP-230](https://jira.tacc.utexas.edu/browse/TUP-230)
- mimics https://github.com/TACC/Core-CMS/pull/498 but on v3.6.0
- ~~https://github.com/TACC/Core-CMS-Resources/pull/144~~ https://github.com/TACC/Core-CMS-Resources/pull/145

## Changes

- pin core-styles to version compatible with CMS v3.6.0
- install `django-recaptcha`
- replace `djangocms-forms` with `djangocms-forms-maintained`
- isolate and document form plugin settings
- copy `c-button` from core-styles v0.5+
- (minimally) overwrite `c-button`
- style form elements (generic and plugin)
- load jQuery with AJAX (to support form)
- sample code to customize captcha
- override form template to:
    1. load css
    2. fix console error
- add required url entry
- add frontera colors for primary button

## Testing

Deployed on [Frontera Pre-Prod](https://pprd.frontera-portal.tacc.utexas.edu/).
Tests on [TUP-230 test page](https://pprd.frontera-portal.tacc.utexas.edu/test/tup-230-form-plugin/). (Must log in.)

1. Create form plugin instance.
    - Complete _required_ settings (no e-mail).[^1]
    - Check "Save to database".
3. Submit form—see success message.[^2]
4. Submit form again—see success message.[^2]
5. Edit form plugin instance.
    - Add SPAM PROTECTION[^3]
6. Submit form—see success message.[^2]
7. In "Site administration", open "Form submissions".
8. Confirm your submissions are there.

[^1]: E-mail feature is delayed until I have bandwidth to research it and configure or fix it.
[^2]: The success message is top center of screen white on black (a `.cms_message`). These should be styled better.
[^3]: There should only be options "None" and "Recaptcha", chose "Recaptcha"

## Screenshots

<details><summary>Screenshots of Fields & Errors</summary>

| Fields | Errors |
| - | - |
| ![fields](https://user-images.githubusercontent.com/62723358/173920489-8e69a45c-e79a-4bdd-b100-6fb3f4cfa15c.jpeg) | ![errors](https://user-images.githubusercontent.com/62723358/173920483-f411a579-0335-4690-abf2-d3e245d50d64.png) |

</details>

<details><summary>Videos of Form Submission with & w/out Captcha</summary>

https://user-images.githubusercontent.com/62723358/173920493-afcdd29f-fddc-4b02-8ac8-ae31cc76b92f.mov

https://user-images.githubusercontent.com/62723358/173920496-99307343-247b-4760-a1bf-c08238de3b80.mov

</details>

## Notes

Authorized developer should add/update a Recaptcha account with domain/s and add secret key to frontera server.
1. [Confluence: **Google reCAPTCHA**, Analytics & Site Verification](https://confluence.tacc.utexas.edu/x/PQCCB)
3. [TACC/Core-Portal-Deployments Frontera CMS captcha settings](https://github.com/TACC/Core-Portal-Deployments/blob/bb88d5b/frontera-portal/camino/pprd.cms.settings_local.py#L13-L15)
